core/src/apps/bitcoin/sign_tx/decred.py

```diff
@@ -208,6 +232,54 @@ def write_tx_output(
             writers.write_uint16(w, DECRED_SCRIPT_VERSION)
         writers.write_bytes_prefixed(w, script_pubkey)
 
+    def validate_sstx_submission(self, txo: TxOutput) -> bytearray:
+        assert txo.address is not None
+        return scripts_decred.output_script_sstxsubmissionpkh(txo.address)
+
+    def validate_sstx_commitment_owned(self, txo: TxOutput) -> bytearray:
+        assert txo.op_return_data is not None  # checked in sanitize_tx_output
+        # Verify that the address this script pays to is owned by the wallet.
+        wantpkh, wantAmt = scripts_decred.pkh_from_sstxcommitment(txo.op_return_data)
+        if wantAmt != self.approver.total_in:
+            raise wire.DataError("sstxcommitment has incorrect amount")
+        key = self.keychain.derive(txo.address_n)
+        gotpkh = ecdsa_hash_pubkey(key.public_key(), self.coin)
+        if gotpkh != wantpkh:
+            raise wire.DataError("sstxcommitment does not pay to this wallet")
```

Comment:
Firstly, I think we should also do some check of the `address_n` path to ensure that it belongs to the same account and that `address_index` is not too large. I'd say it should satisfy the conditions for being a change-address.

Secondly, now that I look at this, when there is only one commitment, it would seem to make more sense to construct the `op_return_data` rather than parse it and verify that it is correct. Actually, even if we decide to support multiple commitments in the future, then it would still make sense to construct the data, except that the amounts would need to be specified in the `TxOutput` messages. It's just another scriptPubKey. Trezor doesn't take any other scriptPubKeys from outside to verify them, instead it constructs them, so there is no reason why this shouldn't be the same.

Reply:
Sounds ok to me. So just pass address_n?

Reply:
I implemented it in this commit: https://github.com/trezor/trezor-firmware/pull/1249/commits/ad3d97a697504c1cb4634c600f571cc789acc71d.
The only thing that I am not sure about is whether the amount field should be filled for commitment outputs. The result is an OP_RETURN, so I eventually have to set it to 0 [here](https://github.com/trezor/trezor-firmware/pull/1249/commits/ad3d97a697504c1cb4634c600f571cc789acc71d#diff-9b0b5241937f81cc834949cd07530935f7e369b18dcc6a52d6525ce842e14671R247), which is a bit of a hack. On the other hand, technically there can be multiple commitment outputs, indeed you originally also allowed a fourth output, which was an external commitment output if I recall correctly. In that case we would need the amount field to tell us how the amounts are distributed. @JoeGruffins, what exactly is the use for external commitment outputs? Is it used for some sort of fee? Do you expect supporting multiple commitment outputs will be needed at some point in Trezor?

Reply:
Yesh, it was used as a fee, but now we pay the fee with a regular transaction afterwards. We do not expect to use multiple commitment outputs in the near future.

Reply:
@matejcik any opinion on the above?
> The only thing that I am not sure about is whether the amount field should be filled for commitment outputs. The result is an OP_RETURN, so I eventually have to set it to 0 [here](https://github.com/trezor/trezor-firmware/pull/1249/commits/ad3d97a697504c1cb4634c600f571cc789acc71d#diff-9b0b5241937f81cc834949cd07530935f7e369b18dcc6a52d6525ce842e14671R247), which is a bit of a hack.